Leader Stats API

gbbafna opened the pull request:
Signed-off-by: Gaurav Bafna <gbbafna@amazon.com>

### Description
Leader Side Status API
 
### Issues Resolved
https://github.com/opensearch-project/cross-cluster-replication/issues/121
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

```
curl  -XGET "${LEADER}/_plugins/_replication/leader_stats"
{
  "num_replicated_indices" : 2,
  "index_stats" : {
    "leader-02" : {
      "operations_read" : 3,
      "translog_size_bytes" : 265,
      "operations_read_lucene" : 0,
      "operations_read_translog" : 3,
      "total_read_time_lucene_millis" : 0,
      "total_read_time_translog_millis" : 9,
      "bytes_read" : 183
    },
    "leader-01" : {
      "operations_read" : 3,
      "translog_size_bytes" : 274,
      "operations_read_lucene" : 0,
      "operations_read_translog" : 3,
      "total_read_time_lucene_millis" : 0,
      "total_read_time_translog_millis" : 3,
      "bytes_read" : 192
    }
  }
```

Same tried on follower side

```
curl  -XGET "${FOLLOWER}/_plugins/_replication/leader_stats"
{
}
```

### ToDo

Integ Tests , Unit Tests

krishna-ggk commented on src/main/kotlin/org/opensearch/replication/action/changes/TransportGetChangesAction.kt:
```diff
@@ -112,6 +123,23 @@ class TransportGetChangesAction @Inject constructor(threadPool: ThreadPool, clus
                         }
                     }
                 }
+
+                val tookInNanos = System.nanoTime() - relativeStartNanos
+                val tookInMillis = TimeUnit.NANOSECONDS.toMillis(tookInNanos)
+                if (fetchFromTranslog) {
+                    indexMetric.latencyTlog.addAndGet(tookInMillis)
+                    indexMetric.opsTlog.addAndGet(ops.size.toLong())
+                } else {
+                    indexMetric.latencyLucene.addAndGet(tookInMillis)
+                    indexMetric.opsLucene.addAndGet(ops.size.toLong())
+                }
```
Actually just noticed that there is fallback possibility after attempting translog. Should we locate the code directly within the if blocks?

gbbafna replied:
Good catch. updated `relativeStartNanos` in that case.  Inclined  to keep code here so that all stats code remain at the same place.

krishna-ggk commented on src/main/kotlin/org/opensearch/replication/action/stats/TransportLeaderStatsAction.kt:
```diff
@@ -0,0 +1,100 @@
+
+    }
+
+     private fun isIndexReplicationActive(index :String) :Boolean {
+        val indicesRouting = clusterService.state().routingTable.indicesRouting
+         if (!indicesRouting.containsKey(index)) {
+             return false
+         }
+
+         return true
+         // Retention lease is the most accurate method to find out active replication, however it blocks the generic thread pool
+         // given that nodeOperation doesn't support async calls
```
Yeah, we should create an issue to track. One option could be to also persist state on leader side on replication starting and cleaning on stopping.

gbbafna replied:
Created one : https://github.com/opensearch-project/cross-cluster-replication/issues/124

saikaranam-amazon replied:
Just saw this - `indexShard` has `getRetentionLeases`  method that can be used to filter the retention leases instead of a transport call. we can discuss

krishna-ggk commented on src/main/kotlin/org/opensearch/replication/action/stats/TransportLeaderStatsAction.kt:
```diff
@@ -0,0 +1,100 @@
+
+       return LeaderNodeStatsResponse(this.clusterService.localNode(), activeStats)
+    }
+}
```
Can you ensure all files have correct line endings?

gbbafna replied:
Fixed

krishna-ggk commented on src/main/kotlin/org/opensearch/replication/task/index/IndexReplicationTask.kt:
```diff
@@ -399,9 +399,7 @@ open class IndexReplicationTask(id: Long, type: String, action: String, descript
-                        if (setting.isPrivateIndex) {
-                            log.info("private setting $key skipping")
-                        } else {
+                        if (!setting.isPrivateIndex) {
```
Wouldn't the visibility help?

gbbafna replied:
This doesn't add much value and was left over log by oversight.

krishna-ggk commented on src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationTask.kt:
```diff
@@ -11,20 +11,6 @@
-import org.opensearch.replication.util.indicesService
-import org.opensearch.replication.util.stackTraceToString
-import org.opensearch.replication.util.suspendExecuteWithRetries
 import kotlinx.coroutines.CancellationException
```
Can we skip this file if it is just rearranging imports?

krishna-ggk commented on src/main/kotlin/org/opensearch/replication/seqno/RemoteClusterStats.kt:
```diff
@@ -0,0 +1,92 @@
+        newStat.bytesRead += stat.bytesRead
+
+        return newStat
+    }
```
I was refering to add method as the only way to add stats. This would have meant even in a place where I need to add bytesRead, I'd have had ton construct Stats object.

However latest rev is fine as we are individually adding the elements.

gbbafna replied:
Yes. This is used only while collecting and adding stats from all the nodes .

saikaranam-amazon commented on src/main/kotlin/org/opensearch/replication/action/changes/TransportGetChangesAction.kt:
```diff
@@ -69,6 +72,13 @@ class TransportGetChangesAction @Inject constructor(threadPool: ThreadPool, clus
+                if (remoteStatsService.stats[shardId.index.name] == null) {
+                    remoteStatsService.stats[shardId.index.name] = RemoteIndexMetric()
+                }
+                val indexMetric = remoteStatsService.stats[shardId.index.name]!!
```
Can we instead perform getOrDefault()?

saikaranam-amazon commented on src/main/kotlin/org/opensearch/replication/action/stats/LeaderNodeStatsResponse.kt:
```diff
@@ -0,0 +1,38 @@
+        super.writeTo(out)
+        out.writeMap(remoteStats, StreamOutput::writeString) { o, v -> v.writeTo(o)}
+    }
+}
```
new line

saikaranam-amazon commented on src/main/kotlin/org/opensearch/replication/action/stats/LeaderStatsRequest.kt:
```diff
@@ -0,0 +1,42 @@
+     * @param in A stream input object.
+     * @throws IOException if the stream cannot be deserialized.
+     */
+    constructor(`in`: StreamInput) : super(`in`)
```
nit: we can instead not use reserved keywords - maybe `inp`
Have to change in below file as well

gbbafna commented:
> Just saw this - indexShard has getRetentionLeases method that can be used to filter the retention leases instead of a transport call. we can discuss

Thanks Sai. This is what we needed . I have implemented that now and moved `Stats` to shard level for the same.

krishna-ggk commented on src/main/kotlin/org/opensearch/replication/action/stats/TransportLeaderStatsAction.kt:
```diff
@@ -0,0 +1,91 @@
+         val indexService = indicesService.indexService(shardId.index) ?: return false
+         val indexShard = indexService.getShard(shardId.id) ?: return false
+
+         var leaseExist = false
+         val retentionLeases = indexShard.getRetentionLeases().leases()
+         for (retentionLease in retentionLeases) {
+             if (retentionLease.id().startsWith(RETENTION_LEASE_PREFIX)) {
+                 leaseExist = true
+             }
+         }
+
+         return leaseExist
+    }
```
Nit - can we return as soon as lease is found?
